### PR TITLE
android: gradle build

### LINKED ISSUE #8460: SDK Tools, Revision 25.3.0 and OpenCVDetectAndroidSDK.cmake
##### System information (version)

- OpenCV => 3.2.0
- Operating System / Platform => Gentoo / Android
- Compiler => Android NDK toolchain gcc

##### Detailed description
OpenCVDetectAndroidSDK.cmake doesn't detect an Android SDK after updating to the SDK Tools, Revision 25.3.0.
```bash
--     SDK target:                  android_sdk_target_status-NOTFOUND
--     android tool:                NO
```
```bash
~/Android/Sdk/tools/android
The android command is no longer available.
For manual SDK and AVD management, please use Android Studio.
For command-line tools, use tools/bin/sdkmanager and tools/bin/avdmanager
```
An older version of the tools just werks.

https://developer.android.com/studio/releases/sdk-tools.html
> SDK Tools, Revision 25.3.0 (March 2017)
>  ...
> Obsolete/deprecated tools have been removed:
> android
> ddms (instead see Using DDMS)
> draw9patch (instead see Draw 9-patch)
> hierarchyviewer (instead see Profile Your Layout with Hierarchy Viewer)
> traceview (instead see Profiling with Traceview and dmtracedump)
> ant scripts
> Project and activity templates

##### Steps to reproduce

1. Update SDK tools using an Android Studio
2. Try to build an OpenCV for Android with Java wrappers and examples support.

---
***alalek***:

Current solution is to downgrade SDK Tools to the latest available version prior 25.3.0: [25.2.5](https://developer.android.com/studio/releases/sdk-tools.html). Direct download links are:
- https://dl.google.com/android/repository/tools_r25.2.5-linux.zip (265Mb)
- https://dl.google.com/android/repository/tools_r25.2.5-macosx.zip (191Mb)
- https://dl.google.com/android/repository/tools_r25.2.5-windows.zip (293Mb)

### PR DESCRIPTION
**Pull request by alalek**
resolves #8460

Usage (build Android package):
```
python2 <opencv_dir>/platforms/android/build_sdk.py --config ndk-17.config.py
```

Android package for testing is [here](http://pullrequest.opencv.org/buildbot/export/opencv_releases/precommit_pack_android/20181219-110522--10816/).

TODO:
- [x] Update infrastructure
  - [x] repaired test_cmake_build
  - [x] added test_gradle (require `test_gradle=1` property)
- [ ] (**postponed**) Repair Android Java tests
- [x] Fix samples build within "make all" target
- [ ] (**postponed**) OpenCV manager => OpenCV loader (debug only) or drop completelly?
  - [ ] new API to replace "initDebug()"
- [ ] (**postponed**) update samples menu implementation ("context" menu is available in android 8 only - min version is 21/16)

```
test_modules=none
buildworker:Android pack=linux-4
docker_image:Android pack=android-gradle
buildworker:Android armeabi-v7a=linux-4
docker_image:Android armeabi-v7a=android-gradle
test_gradle=1
allow_multiple_commits=1
```

### CODE REVIEW

File: platforms/android/build_sdk.py [hunk 1]
```diff
@@ -303,7 +236,6 @@ def gather_results(self, engines):
     parser.add_argument('--sign_with', help="Certificate to sign the Manager apk")
```
Comment (crazyhappygame) [inline platforms/android/build_sdk.py]: Could you change 
parser.add_argument('--config', default='ndk-10.config.py',
to?
parser.add_argument('--config', default='ndk-17.config.py',

File: platforms/android/build_sdk.py [hunk 2]
```diff
@@ -303,7 +236,6 @@ def gather_results(self, engines):
...
     parser.add_argument('--build_doc', action="store_true", help="Build javadoc")
     parser.add_argument('--no_ccache', action="store_true", help="Do not use ccache during library build")
```
Comment (crazyhappygame) [inline platforms/android/build_sdk.py]: Could you disable ccache by default? (maybe change no_ccache to enable_ccache)
ccache in not available by default on Linux
ccache is not available on Windows

**Comment by crazyhappygame** [comment]
@alalek 
1. When will you merge this PR? 
2. When will you create clang build for android (opencv 3.4.4, 4.0) and add to https://opencv.org/releases.html ?

**Comment by alalek** [comment]
You can try Android package with this PR from here (see GitHub)

**Comment by alalek** [comment]
:+1:

**Pull request merged**